[SPARK-32976][SQL]Support column list in INSERT statement

yaooqinn opened the pull request:
### What changes were proposed in this pull request?
 
#### JIRA expectations
```
   INSERT currently does not support named column lists.  
   
   INSERT INTO <table> (col1, col2,…) VALUES( 'val1', 'val2', … )
   Note, we assume the column list contains all the column names. Issue an exception if the list is not complete. The column order could be different from the column order defined in the table definition.    
```
#### implemetations
In this PR, we add a column list  as an optional part to the `INSERT OVERWRITE/INTO` statements:
```
  /**
   * {{{
   *   INSERT OVERWRITE TABLE tableIdentifier [partitionSpec [IF NOT EXISTS]]? [identifierList] ...
   *   INSERT INTO [TABLE] tableIdentifier [partitionSpec]  [identifierList] ...
   * }}}
   */
```
The column list represents all expected columns with an explicit order that you want to insert to the target table. **Particularly**,  we assume the column list contains all the column names in the current implementation, it will fail when the list is incomplete.

In **Analyzer**, we add a code path to resolve the column list in the `ResolveOutputRelation` rule before it is transformed to v1 or v2 command. It will fail here if the list has any field that not belongs to the target table.

Then, for v2 command, e.g. `AppendData`, we use the resolved column list and output of the target table to resolve the output of the source query `ResolveOutputRelation` rule. If the list has duplicated columns, we fail. If the list is not empty but the list size does not match the target table, we fail. If no other exceptions occur, we use the column list to map the output of the source query to the output of the target table.  The column list will be set to Nil and it will not hit the rule again after it is resolved.

for v1 command, those all happen in the `PreprocessTableInsertion` rule





### Why are the changes needed?
 new feature support

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

yes, insert into/overwrite table support specify column list
### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
new tests

yaooqinn committed:
[WIP][SPARK-32976][SQL]Support column list in INSERT statement

yaooqinn committed:
Merge branch 'master' into SPARK-32976

SparkQA commented:
**[Test build #129185 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129185/testReport)** for PR 29893 at commit [`99056c4`](https://github.com/apache/spark/commit/99056c47af4376467d1117ee0ef5018c6bf8985c).

yaooqinn committed:
add more tests

SparkQA commented:
**[Test build #129185 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129185/testReport)** for PR 29893 at commit [`99056c4`](https://github.com/apache/spark/commit/99056c47af4376467d1117ee0ef5018c6bf8985c).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129185/
Test FAILed.

SparkQA commented:
**[Test build #129186 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129186/testReport)** for PR 29893 at commit [`c0643f6`](https://github.com/apache/spark/commit/c0643f6cbd24b561e079172d1d8953ba9af9937d).

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33800/
Test FAILed.

SparkQA commented:
**[Test build #129186 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129186/testReport)** for PR 29893 at commit [`c0643f6`](https://github.com/apache/spark/commit/c0643f6cbd24b561e079172d1d8953ba9af9937d).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129186/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33802/
Test FAILed.

yaooqinn committed:
build

SparkQA commented:
**[Test build #129203 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129203/testReport)** for PR 29893 at commit [`40e2ffd`](https://github.com/apache/spark/commit/40e2ffd43ff7c41dcf6c162daf0d9bc672a86996).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33818/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33818/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33818/
Test FAILed.

SparkQA commented:
**[Test build #129203 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129203/testReport)** for PR 29893 at commit [`40e2ffd`](https://github.com/apache/spark/commit/40e2ffd43ff7c41dcf6c162daf0d9bc672a86996).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129203/
Test FAILed.

yaooqinn committed:
fix ut

SparkQA commented:
**[Test build #129221 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129221/testReport)** for PR 29893 at commit [`e756a55`](https://github.com/apache/spark/commit/e756a558d89a077108d50103af441dffa4ba4263).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33837/

yaooqinn committed:
style

SparkQA commented:
**[Test build #129229 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129229/testReport)** for PR 29893 at commit [`2b164ed`](https://github.com/apache/spark/commit/2b164eddd80145e9104546a7e7b41d340fd9481f).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33837/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33837/
Test PASSed.

yaooqinn committed:
V2 Support

SparkQA commented:
**[Test build #129232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129232/testReport)** for PR 29893 at commit [`086cfa8`](https://github.com/apache/spark/commit/086cfa8260b1aba1bab774c80f8c3ff5c6d2f31a).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33845/

yaooqinn committed:
add v2 test

SparkQA commented:
**[Test build #129233 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129233/testReport)** for PR 29893 at commit [`3a02681`](https://github.com/apache/spark/commit/3a0268107996abf3bae6a969181b269b1fc19481).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33845/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33845/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33848/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33849/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33848/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33848/
Test FAILed.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33849/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33849/
Test FAILed.

SparkQA commented:
**[Test build #129221 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129221/testReport)** for PR 29893 at commit [`e756a55`](https://github.com/apache/spark/commit/e756a558d89a077108d50103af441dffa4ba4263).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129221/
Test PASSed.

SparkQA commented:
**[Test build #129229 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129229/testReport)** for PR 29893 at commit [`2b164ed`](https://github.com/apache/spark/commit/2b164eddd80145e9104546a7e7b41d340fd9481f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129229/
Test PASSed.

SparkQA commented:
**[Test build #129232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129232/testReport)** for PR 29893 at commit [`086cfa8`](https://github.com/apache/spark/commit/086cfa8260b1aba1bab774c80f8c3ff5c6d2f31a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129232/
Test PASSed.

SparkQA commented:
**[Test build #129233 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129233/testReport)** for PR 29893 at commit [`3a02681`](https://github.com/apache/spark/commit/3a0268107996abf3bae6a969181b269b1fc19481).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129233/
Test PASSed.

yaooqinn committed:
fix and add v2 test

SparkQA commented:
**[Test build #129265 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129265/testReport)** for PR 29893 at commit [`4af3a63`](https://github.com/apache/spark/commit/4af3a63aee8a2526f36b42702bcae85b18b44a9a).

yaooqinn committed:
nit

SparkQA commented:
**[Test build #129266 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129266/testReport)** for PR 29893 at commit [`ccd11f3`](https://github.com/apache/spark/commit/ccd11f36a1cbc32d9f123e8a821bd0d2169ca949).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33882/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33883/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33882/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33882/
Test PASSed.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33883/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33883/
Test PASSed.

SparkQA commented:
**[Test build #129266 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129266/testReport)** for PR 29893 at commit [`ccd11f3`](https://github.com/apache/spark/commit/ccd11f36a1cbc32d9f123e8a821bd0d2169ca949).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #129265 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129265/testReport)** for PR 29893 at commit [`4af3a63`](https://github.com/apache/spark/commit/4af3a63aee8a2526f36b42702bcae85b18b44a9a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129266/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129265/
Test PASSed.

yaooqinn committed:
more test

yaooqinn commented:
cc @gatorsmile @cloud-fan @wangyum thanks

SparkQA commented:
**[Test build #129299 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129299/testReport)** for PR 29893 at commit [`a76c9ba`](https://github.com/apache/spark/commit/a76c9ba5e2c41b9aa68b6a5e170f7d2750af7514).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33915/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/33915/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/33915/
Test PASSed.

gatorsmile commented:
Thanks for working on it! Happy Holiday!

SparkQA commented:
**[Test build #129299 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129299/testReport)** for PR 29893 at commit [`a76c9ba`](https://github.com/apache/spark/commit/a76c9ba5e2c41b9aa68b6a5e170f7d2750af7514).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129299/
Test PASSed.

cloud-fan commented:
can we enrich the PR description to explain how this feature is implemented in the analyzer phase? e.g. how we resolve logical plans for insertion.

wangyum commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -337,8 +337,8 @@ query
-    : INSERT OVERWRITE TABLE? multipartIdentifier (partitionSpec (IF NOT EXISTS)?)?                         #insertOverwriteTable
-    | INSERT INTO TABLE? multipartIdentifier partitionSpec? (IF NOT EXISTS)?                                #insertIntoTable
+    : INSERT OVERWRITE TABLE? multipartIdentifier identifierList?  (partitionSpec (IF NOT EXISTS)?)?        #insertOverwriteTable
+    | INSERT INTO TABLE? multipartIdentifier identifierList?  partitionSpec? (IF NOT EXISTS)?               #insertIntoTable
```
Move `identifierList?` to the end? This is the order of Hive:
```sql
hive> SELECT version();
2.3.7 rcb213d88304034393d68cc31a95be24f5aac62b6
hive> CREATE TABLE spark_32976(id int, name string) PARTITIONED BY (part string);
hive> INSERT INTO spark_32976 PARTITION (part = '1') (name) SELECT 'name1';
hive> SELECT * FROM spark_32976;
NULL	name1	1
```

yaooqinn committed:
address comments

yaooqinn commented:
Thanks for reviewing, sorry for the late reply as I spent a long holiday with my family. I've addressed the comment, please check again @gatorsmile @cloud-fan @wangyum thanks.

SparkQA commented:
**[Test build #129666 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129666/testReport)** for PR 29893 at commit [`c76a2aa`](https://github.com/apache/spark/commit/c76a2aacb86411ef9d0c58ac996b274dce49beb7).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34270/

SparkQA commented:
**[Test build #129666 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129666/testReport)** for PR 29893 at commit [`c76a2aa`](https://github.com/apache/spark/commit/c76a2aacb86411ef9d0c58ac996b274dce49beb7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129666/
Test FAILed.

yaooqinn committed:
fix ut

SparkQA commented:
**[Test build #129673 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129673/testReport)** for PR 29893 at commit [`f7e8d18`](https://github.com/apache/spark/commit/f7e8d186320aa805d679942937b12b3a5141fb89).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34270/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34270/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34277/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34277/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34277/
Test FAILed.

SparkQA commented:
**[Test build #129673 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129673/testReport)** for PR 29893 at commit [`f7e8d18`](https://github.com/apache/spark/commit/f7e8d186320aa805d679942937b12b3a5141fb89).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129673/
Test PASSed.

cloud-fan commented:
Sorry for the late review. I have a simpler idea about how to implement it:
1. Add a new field `userSpecifiedColums` to `InsertIntoStatement`.
2. Add a new analyzer rule to resolve `InsertIntoStatement` whose `userSpecifiedColums` is not Nil, by adding a project to the input query to reorder the columns. This rule will set `userSpecifiedColums` to Nil at the end, so that it's idempotent.
3. Other rules that match `InsertIntoStatement` should wait until `userSpecifiedColums` is Nil.

Thus we don't need to change the v2 write commands.

gatorsmile commented:
ping @yaooqinn  Could you help address the above comments? This is a nice feature we should have for ANSI SQL

yaooqinn commented:
HI @cloud-fan @gatorsmile,  I will focus on this next week.

yaooqinn committed:
Merge branch 'master' into SPARK-32976

SparkQA commented:
**[Test build #131661 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131661/testReport)** for PR 29893 at commit [`c43b964`](https://github.com/apache/spark/commit/c43b96404cbdfeb859784304f2174fc28f66b357).

SparkQA commented:
**[Test build #131661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131661/testReport)** for PR 29893 at commit [`c43b964`](https://github.com/apache/spark/commit/c43b96404cbdfeb859784304f2174fc28f66b357).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131661/

yaooqinn committed:
fix tests

SparkQA commented:
**[Test build #131673 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131673/testReport)** for PR 29893 at commit [`c74d4d7`](https://github.com/apache/spark/commit/c74d4d7a0e0768fa5d9970a9e7ddd109306d6caa).

yaooqinn committed:
add v1 support

SparkQA commented:
**[Test build #131673 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131673/testReport)** for PR 29893 at commit [`c74d4d7`](https://github.com/apache/spark/commit/c74d4d7a0e0768fa5d9970a9e7ddd109306d6caa).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

yaooqinn committed:
nit

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131673/

SparkQA commented:
**[Test build #131678 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131678/testReport)** for PR 29893 at commit [`475e790`](https://github.com/apache/spark/commit/475e790f7eeae80c4123f93c74952eb24f056897).

SparkQA commented:
**[Test build #131678 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131678/testReport)** for PR 29893 at commit [`475e790`](https://github.com/apache/spark/commit/475e790f7eeae80c4123f93c74952eb24f056897).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131678/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36262/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36273/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36278/

yaooqinn commented:
retest this please

yaooqinn commented:
retest this please

SparkQA commented:
**[Test build #131750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131750/testReport)** for PR 29893 at commit [`475e790`](https://github.com/apache/spark/commit/475e790f7eeae80c4123f93c74952eb24f056897).

SparkQA commented:
**[Test build #131750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131750/testReport)** for PR 29893 at commit [`475e790`](https://github.com/apache/spark/commit/475e790f7eeae80c4123f93c74952eb24f056897).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131750/

yaooqinn commented:
retest this please

SparkQA commented:
**[Test build #131766 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131766/testReport)** for PR 29893 at commit [`475e790`](https://github.com/apache/spark/commit/475e790f7eeae80c4123f93c74952eb24f056897).

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36348/

SparkQA commented:
**[Test build #131766 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131766/testReport)** for PR 29893 at commit [`475e790`](https://github.com/apache/spark/commit/475e790f7eeae80c4123f93c74952eb24f056897).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131766/

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -299,13 +301,16 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
       ctx: InsertIntoTableContext): InsertTableParams = withOrigin(ctx) {
     val tableIdent = visitMultipartIdentifier(ctx.multipartIdentifier)
+    val cols = Option(ctx.identifierList())
+      .map(visitIdentifierList)
```
Can we add a check that column name must be one part?

yaooqinn replied:
unnecessary, it comes from a grammar that accepts a one-part name

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala:
```diff
@@ -290,6 +291,7 @@ case class DropViewStatement(
 case class InsertIntoStatement(
     table: LogicalPlan,
     partitionSpec: Map[String, Option[String]],
+    userSpecifiedCols: Seq[Attribute],
```
I'm not sure if this should be `Attribute` not `String`. `InsertIntoStatement`'s child is `query`, not `table`, which means these attributes will be resolved based on `query`, while they should be resolved by `table`.

yaooqinn replied:
ah, `String` here is much simpler and more rational.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -3718,6 +3718,49 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     }
   }
+
+  test("SPARK-32976: Support column list specification in insert into statement") {
```
can we create a `insert.sql` test file and move this test there?

yaooqinn replied:
Based on this requirement, I create the base trait and 3 concrete test suites, but w/ `V1SQLInsertTestSuite` trait for bypassing different error handling between v2/v1 tables, which I guess is not the right thing to do in this PR.

FYI, when the final query's output mismatches w/ the target table output, v1 throws errors at org.apache.spark.sql.execution.datasources.DataSourceAnalysis#convertStaticPartitions, while v2 at org.apache.spark.sql.catalyst.analysis.TableOutputResolver#resolveOutputColumns

To merge these two parts rashly would introduce a large number of LOC in other test cases and may
 make this PR hard to review.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -3104,6 +3106,66 @@ class Analyzer(override val catalogManager: CatalogManager)
+
+    private def resolveUserSpecifiedColumns(i: InsertIntoStatement): Seq[Attribute] = {
+      val resolved = i.userSpecifiedCols.map {
+        case u: UnresolvedAttribute => withPosition(u) {
```
Does other databases support `INSERT INTO t(t.col) ...`? Otherwise using `String` instead of `UnresolvedAttribute` is simpler.

yaooqinn replied:
yea, it's wired. the column name should have only one part

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36364/

yaooqinn commented:
I will address the comment for the main code change first but leave the test part tomorrow. It’s too late now :)

yaooqinn committed:
attribute -> string

SparkQA commented:
**[Test build #131796 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131796/testReport)** for PR 29893 at commit [`86d0032`](https://github.com/apache/spark/commit/86d0032ad00f7a1f10e1963070e39a24e640998d).

SparkQA commented:
**[Test build #131796 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131796/testReport)** for PR 29893 at commit [`86d0032`](https://github.com/apache/spark/commit/86d0032ad00f7a1f10e1963070e39a24e640998d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131796/

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36394/

yaooqinn committed:
nit

SparkQA commented:
**[Test build #131826 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131826/testReport)** for PR 29893 at commit [`713b5b9`](https://github.com/apache/spark/commit/713b5b9a1d47ca639f36b690a2435363e9eb4fff).

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36424/

SparkQA commented:
**[Test build #131826 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131826/testReport)** for PR 29893 at commit [`713b5b9`](https://github.com/apache/spark/commit/713b5b9a1d47ca639f36b690a2435363e9eb4fff).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131826/

yaooqinn committed:
SQLInsertTestSuite

SparkQA commented:
**[Test build #131844 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131844/testReport)** for PR 29893 at commit [`3afb4c0`](https://github.com/apache/spark/commit/3afb4c0f92fc9f549d3dbd962c263518a65f6915).

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36441/

SparkQA commented:
**[Test build #131844 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131844/testReport)** for PR 29893 at commit [`3afb4c0`](https://github.com/apache/spark/commit/3afb4c0f92fc9f549d3dbd962c263518a65f6915).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131844/

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -3104,6 +3106,62 @@ class Analyzer(override val catalogManager: CatalogManager)
+          TableOutputResolver.checkField(
+            tableAttr, nameToQueryExpr(tableAttr), byName = false, conf, err => errors += err)
+        } else {
+          None
```
when can we go to this branch?

yaooqinn replied:
this is for static partition columns

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala:
```diff
@@ -522,7 +523,8 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
         } else {
-          OverwriteByExpression.byPosition(table, df.logicalPlan, Literal(true), extraOptions.toMap)
+          OverwriteByExpression.byPosition(
+            table, df.logicalPlan, Literal(true), extraOptions.toMap)
```
unnecessry change.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,244 @@
+    } else ""
+    withTempView(tmpView) {
+      insert.createOrReplaceTempView(tmpView)
+      val overwrite = if (mode == SaveMode.Overwrite) "OVERWRITE" else "INTO"
```
It's weird to see `SaveMode` in SQL test. Can we use `overwrite: Boolean`?

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  test("insert with column list - follow table output order") {
+    val t1 = "t1"
```
we can hardcode table name if it's very simple like `t1`.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,244 @@
+    withTable(t1) {
+      val df = Seq((1, 2L, "3")).toDF()
+      val cols = Seq("c1", "c2", "c3")
+      createTable(t1, cols, Seq("int", "long", "string"))
```
it's more readable if we pass schema string directly `createTable(t1, "c1 int, c2 long, c3 string")`

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,244 @@
+    val cols = Seq("c1", "c2", "c3", "c4")
+    val df = Seq((1, 2, 3, 4)).toDF(cols: _*)
+    withTable(t1) {
+      createTable(t1, cols, Seq("int", "int", "int", "int"), cols.takeRight(2))
```
we should keep each test case small. This should be a new test case: `insert with column list - follow table output order + partitioned table`

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,244 @@
+      }
+    }
+
+    withTable(t1) {
```
ditto, `insert with column list - follow table output order + partitioned table and write with static partitions`

yaooqinn committed:
nit

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,244 @@
+    withTable(table) {
+      sql(s"CREATE TABLE $table(c1 INT, c2 INT, c3 INT) USING $format")
+      val e1 = intercept[AnalysisException](sql(s"INSERT INTO $table (c1) values(1)"))
+      assert(e1.getMessage.contains("target table has 3 column(s) but the inserted data has 1"))
```
only error message is different? then we can put test in the base trait
```
val e = intercept[AnalysisException]...
assert(e.message.contains(DSV1Msg) || e.message.contains(DSV2Msg))
```

SparkQA commented:
**[Test build #131868 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131868/testReport)** for PR 29893 at commit [`6f9b42e`](https://github.com/apache/spark/commit/6f9b42e963b46fb8fa049c4c88909d048b5b1b66).

yaooqinn committed:
address comments

SparkQA commented:
**[Test build #131871 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131871/testReport)** for PR 29893 at commit [`7680da2`](https://github.com/apache/spark/commit/7680da271c3884093a95f7352fe6e9067c537430).

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,230 @@
+    val cols = Seq("c1", "c2", "c3", "c4")
+    val df = Seq((1, 2, 3, 4)).toDF(cols: _*)
+    withTable("t1") {
+      createTable("t1", cols, Seq("int", "int", "int", "int"), cols.takeRight(2))
```
this is a partitioned table, should be put in the test  `insert with column list - follow table output order + partitioned table`

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,230 @@
+    val cols = Seq("c1", "c2", "c3", "c4")
+    val df = Seq((1, 2, 3, 4)).toDF(cols: _*)
+    withTable("t1") {
+      createTable("t1", cols, Seq("int", "int", "int", "int"), cols.takeRight(2))
```
ditto

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,230 @@
+    withTable(t1) {
+      val cols = Seq("c1", "c2", "c3")
+      createTable(t1, cols, Seq("int", "long", "string"))
+      val e1 = intercept[AnalysisException](sql(s"INSERT INTO $t1 (c1, c2) values(1, 2, 3)"))
```
how about  `INSERT INTO $t1 (c1, c2) values(1, 2, 3)`

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,230 @@
+    val v2Msg = "Cannot write to 'testcat.t12', not enough data columns:"
+    val cols = Seq("c1", "c2", "c3", "c4")
+
+    withTable("t12") {
```
can we use `t1`?

yaooqinn committed:
address comments

yaooqinn committed:
address comments

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36468/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36465/

SparkQA commented:
**[Test build #131877 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131877/testReport)** for PR 29893 at commit [`833ac41`](https://github.com/apache/spark/commit/833ac416aac680059ce95fffc09c3234e87e0757).

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -3104,6 +3106,48 @@ class Analyzer(override val catalogManager: CatalogManager)
+        query.failAnalysis(
+          s"""Cannot write to table due to mismatched user specified columns and data columns:
+             |Specified columns: ${cols.map(c => s"'${c.name}'").mkString(", ")}
+             |Data columns: ${query.output.map(c => s"'${c.name}'").mkString(", ")}"""
```
The data column names don't need to match the user-specified column names, e.g. `INSERT INTO t(a, b) SELECT 1, 2`. The thing only matters is the number of columns. Can we refine the error message?

yaooqinn replied:
OK, btw, org.apache.spark.sql.catalyst.analysis.TableOutputResolver#resolveOutputColumns seems to have the same issuse?

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,224 @@
+
+  protected def processInsert(
+      tableName: String,
+      insert: DataFrame,
```
nit: `input`

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,224 @@
+    }
+  }
+
+  test("insert with column list - table output reorder  + partitioned table") {
```
nit: remove one space before `+`

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36474/

yaooqinn committed:
address comments

SparkQA commented:
**[Test build #131871 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131871/testReport)** for PR 29893 at commit [`7680da2`](https://github.com/apache/spark/commit/7680da271c3884093a95f7352fe6e9067c537430).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131871/

SparkQA commented:
**[Test build #131880 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131880/testReport)** for PR 29893 at commit [`17d0564`](https://github.com/apache/spark/commit/17d056467751785664b6fd5c89c602f7c3e07e94).

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36477/

SparkQA commented:
**[Test build #131868 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131868/testReport)** for PR 29893 at commit [`6f9b42e`](https://github.com/apache/spark/commit/6f9b42e963b46fb8fa049c4c88909d048b5b1b66).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131868/

SparkQA commented:
**[Test build #131877 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131877/testReport)** for PR 29893 at commit [`833ac41`](https://github.com/apache/spark/commit/833ac416aac680059ce95fffc09c3234e87e0757).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131877/

SparkQA commented:
**[Test build #131880 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131880/testReport)** for PR 29893 at commit [`17d0564`](https://github.com/apache/spark/commit/17d056467751785664b6fd5c89c602f7c3e07e94).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131880/

yaooqinn commented:
retest this please

SparkQA commented:
**[Test build #131889 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131889/testReport)** for PR 29893 at commit [`17d0564`](https://github.com/apache/spark/commit/17d056467751785664b6fd5c89c602f7c3e07e94).

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36486/

SparkQA commented:
**[Test build #131889 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131889/testReport)** for PR 29893 at commit [`17d0564`](https://github.com/apache/spark/commit/17d056467751785664b6fd5c89c602f7c3e07e94).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131889/

cloud-fan commented:
thanks, merging to master!

yaooqinn commented:
thanks for merging. do we need to update the SQL reference doc too?

maropu commented:
Nice feature, @yaooqinn ! It would be nice to update the SQL doc, too, as you said above.

yaooqinn commented:
OK~, I'll make a followup very very soon

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSQLInsertTestSuite.scala:
```diff
@@ -0,0 +1,25 @@
+import org.apache.spark.sql.SQLInsertTestSuite
+import org.apache.spark.sql.hive.test.TestHiveSingleton
+
+class HiveSQLInsertTestSuite extends SQLInsertTestSuite with TestHiveSingleton {
```
We need to set ` spark.conf.set("hive.exec.dynamic.partition.mode", "nonstrict")` before executing this suite; otherwise, `test("insert with column list - follow table output order + partitioned table")` will fail. 

The reason why it does not fail because some test cases [running before this suite] do not change the default value of `hive.exec.dynamic.partition.mode` back to `strict`. However, the order of test suite execution is not deterministic.

gatorsmile replied:
How about something like

```
  protected override def beforeAll(): Unit = {
    spark.conf.set("hive.exec.dynamic.partition.mode", "nonstrict")
    super.beforeAll()
  }

  protected override def afterAll(): Unit = {
    try {
      spark.conf.set("hive.exec.dynamic.partition.mode", "strict")
    } finally {
      super.afterAll()
    }
  }
```

cloud-fan replied:
Good catch! We can follow `HiveCharVarcharTestSuite` to set/restore this conf.

yaooqinn replied:
I have submitted a followup

maropu commented:
@yaooqinn kindly ping: I've filed jira so that we don't forget to do it. https://issues.apache.org/jira/browse/SPARK-33877 `branch-3.1` includes this commit, so I think its better to document it until v3.1.0 released. cc: @HyukjinKwon
